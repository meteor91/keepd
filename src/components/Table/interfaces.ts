import type React from 'react';

export type DataIndex = string | number | ReadonlyArray<(string | number)>;

export interface ColumnType<RecordType> {
    title?: string;
    dataIndex?: DataIndex;
    // dataIndex: string | number;
    render?: (value: any, record: RecordType, index: number) => React.ReactNode;
}

export type GetRowKey<RecordType> = (record: RecordType, index?: number) => React.Key;
export type DefaultRecordType = Record<string, any>;
