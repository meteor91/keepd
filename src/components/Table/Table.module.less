.table {
  border-collapse: collapse;

  .thead {
    .tr {
      .cell {
        background: #F9FAFB;
      }
    }
  }

  .thead,
  .tbody
  {
    .tr {
      .cell {
        padding: 15px 20px 15px;
      }
    }
  }

  .tbody {
    .tr + .tr {
        border-top: #E9EFF6 solid 1px;
    }
  }
}