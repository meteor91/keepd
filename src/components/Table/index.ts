export { Table, type TableProps } from './Table';
export { ColumnDef } from '@tanstack/react-table';