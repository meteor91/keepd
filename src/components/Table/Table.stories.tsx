import React, { useState } from 'react';
import type { Meta, StoryObj } from '@storybook/react';

import { Table, type TableProps } from './Table';
import { type ColumnType } from './interfaces';

const meta: Meta<typeof Table> = {
    title: 'Table',
    component: Table,
};

export default meta;

type Story = StoryObj<typeof Table>;

export const Basic: Story = {
    render: () => <Table dataSource={dataSource} columns={columns}/>,
};

interface ExampleData {
    name: string;
    age: number;
    gamePass: {
        enabled: boolean;
    };
}

const dataSource: ExampleData[] = [
    {
        name: 'Tom',
        age: 27,
        gamePass: {
            enabled: true,
        },
    },
    {
        name: 'Jerry',
        age: 22,
        gamePass: {
            enabled: false,
        },
    },
];

const columns: Array<ColumnType<ExampleData>> = [
    {
        title: 'name',
        dataIndex: 'name',
    },
    {
        title: 'age',
        dataIndex: 'age',
    },
    {
        title: 'game pass',
        dataIndex: 'gamePass',
        render: (value) => value.enabled === true ? 'yes' : 'no',
    },
];
