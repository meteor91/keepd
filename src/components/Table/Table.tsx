import React from 'react';
// import clsx from 'clsx';
import { type ColumnType, type GetRowKey } from './interfaces';
import style from './Table.module.less';
import { BodyRow } from './components/BodyRow';

export interface TableProps<RecordType = unknown> {
    dataSource: RecordType[];
    columns: Array<ColumnType<RecordType>>;
    rowKey?: string | keyof RecordType | GetRowKey<RecordType>;
}

// const myCom: <T>SFC<Prop<T>> = <T>(props: Prop<T>)=> <div>test</div>

// eslint-disable-next-line @typescript-eslint/comma-dangle
export const Table = <T, >(props: TableProps<T>): React.ReactElement => {
    const { columns, dataSource } = props;
    return (
        <table className={style.table}>
            <thead className={style.thead}>
                <tr>
                    {columns.map((column: ColumnType<T>) => {
                        return (
                            <th key={column.dataIndex as string} className={style.cell}>
                                {column.title}
                            </th>
                        );
                    })}
                </tr>
            </thead>
            <tbody>
                {dataSource.map((rowData: T, index) => {
                    return <BodyRow key={index} rowData={rowData} columns={columns} />;
                })}
            </tbody>
        </table>
    );
};
