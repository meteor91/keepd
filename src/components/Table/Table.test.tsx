import React from 'react';
import { render, type RenderResult } from '@testing-library/react';
import { Table, type TableProps } from './Table';

describe('Table', () => {
    const dataSource = [
        {
            name: 'Tom',
            age: 27,
            gamePass: {
                enabled: true,
            },
        },
        {
            name: 'Jerry',
            age: 22,
            gamePass: {
                enabled: false,
            },
        },
    ];

    const columns = [
        {
            title: 'name',
            dataIndex: 'name',
        },
        {
            title: 'age',
            dataIndex: 'age',
        },
        {
            title: 'game pass',
            dataIndex: 'gamePass',
            render: (value: any) => value.enabled === true ? 'yes' : 'no',
        },
    ];

    const renderTable = (props: TableProps<any>): RenderResult => {
        return render(<Table dataSource={props.dataSource} columns={props.columns} />);
    };

    it('should render correct', () => {
        const { container } = renderTable({
            columns,
            dataSource,
        });

        expect(container.firstChild).toMatchSnapshot();
    });
});
