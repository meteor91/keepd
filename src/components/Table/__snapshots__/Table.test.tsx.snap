// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render correct 1`] = `
<table
  class="table"
>
  <thead
    class="thead"
  >
    <tr
      class="tr"
    >
      <th
        class="cell"
      >
        name
      </th>
      <th
        class="cell"
      >
        age
      </th>
      <th
        class="cell"
      >
        game pass
      </th>
    </tr>
  </thead>
  <tbody
    class="tbody"
  >
    <tr
      class="tr"
    >
      <td
        class="cell"
      >
        Tom
      </td>
      <td
        class="cell"
      >
        27
      </td>
      <td
        class="cell"
      >
        yes
      </td>
    </tr>
    <tr
      class="tr"
    >
      <td
        class="cell"
      >
        Jerry
      </td>
      <td
        class="cell"
      >
        22
      </td>
      <td
        class="cell"
      >
        no
      </td>
    </tr>
  </tbody>
</table>
`;
